### PR TITLE
esp32s3/spi: Add SPI bus init in bringup and fix SPI bus 2 and 3 conflict

### LINKED ISSUE #11539: ESP32S3:  SPI cannot use.
## Problem:  
Use system spitool, the spi bus cannot be got.  
  

## Config:
``` #
# This file is autogenerated: PLEASE DO NOT EDIT IT.
#
# You can use "make menuconfig" to make any modifications to the installed .config file.
# You can then do "make savedefconfig" to generate a new defconfig file that includes your
# modifications.
#
# CONFIG_ARCH_LEDS is not set
# CONFIG_NSH_ARGCAT is not set
# CONFIG_NSH_CMDOPT_HEXDUMP is not set
CONFIG_ARCH="xtensa"
CONFIG_ARCH_BOARD="esp32s3-devkit"
CONFIG_ARCH_BOARD_COMMON=y
CONFIG_ARCH_BOARD_ESP32S3_DEVKIT=y
CONFIG_ARCH_CHIP="esp32s3"
CONFIG_ARCH_CHIP_ESP32S3=y
CONFIG_ARCH_CHIP_ESP32S3WROOM1=y
CONFIG_ARCH_INTERRUPTSTACK=2048
CONFIG_ARCH_STACKDUMP=y
CONFIG_ARCH_XTENSA=y
CONFIG_BOARD_LOOPSPERMSEC=16717
CONFIG_BUILTIN=y
CONFIG_DEBUG_ASSERTIONS=y
CONFIG_DEBUG_FEATURES=y
CONFIG_ESP32S3_I2C0=y
CONFIG_ESP32S3_I2C0_SCLPIN=6
CONFIG_ESP32S3_I2C0_SDAPIN=7
CONFIG_ESP32S3_I2C1=y
CONFIG_ESP32S3_I2C1_SCLPIN=15
CONFIG_ESP32S3_I2C1_SDAPIN=16
CONFIG_ESP32S3_SPI2=y
CONFIG_ESP32S3_SPI2_CLKPIN=2
CONFIG_ESP32S3_SPI2_CSPIN=1
CONFIG_ESP32S3_SPI2_MISOPIN=41
CONFIG_ESP32S3_SPI2_MOSIPIN=42
CONFIG_ESP32S3_SPI3=y
CONFIG_ESP32S3_SPI3_CLKPIN=13
CONFIG_ESP32S3_SPI3_CSPIN=11
CONFIG_ESP32S3_SPI3_MISOPIN=12
CONFIG_ESP32S3_SPI3_MOSIPIN=14
CONFIG_ESP32S3_SPI_DMA=y
CONFIG_ESP32S3_UART0=y
CONFIG_FS_PROCFS=y
CONFIG_HAVE_CXX=y
CONFIG_HAVE_CXXINITIALIZE=y
CONFIG_I2CTOOL_DEFFREQ=100000
CONFIG_I2CTOOL_MAXBUS=1
CONFIG_I2C_RESET=y
CONFIG_IDLETHREAD_STACKSIZE=3072
CONFIG_INIT_ENTRYPOINT="nsh_main"
CONFIG_INTELHEX_BINARY=y
CONFIG_NSH_ARCHINIT=y
CONFIG_NSH_BUILTIN_APPS=y
CONFIG_NSH_FILEIOSIZE=512
CONFIG_NSH_LINELEN=64
CONFIG_NSH_READLINE=y
CONFIG_PREALLOC_TIMERS=4
CONFIG_RAM_SIZE=114688
CONFIG_RAM_START=0x20000000
CONFIG_RR_INTERVAL=200
CONFIG_SCHED_WAITPID=y
CONFIG_SMP=y
CONFIG_SMP_NCPUS=2
CONFIG_SPITOOL_MINBUS=2
CONFIG_START_DAY=28
CONFIG_START_MONTH=6
CONFIG_START_YEAR=2022
CONFIG_SYSLOG_BUFFER=y
CONFIG_SYSTEM_I2CTOOL=y
CONFIG_SYSTEM_NSH=y
CONFIG_SYSTEM_SPITOOL=y
CONFIG_UART0_SERIAL_CONSOLE=y
```

## Hardware :
Module: ESP32-S3-WROOM-1(M0N4R2)

### PR DESCRIPTION
**Pull request by w2016561536**
## Summary
1. Add spi bus init in esp32s3_bringup.c
2. Fix IOMUX conflict between spi bus 2 and 3
3. Add spi defconfig

## Impact
Fix esp32s3 spi bus and close #11539 

## Testing
QQ截图20240114223540
Spi bus 2 and 3 mount icm-20602 and FM25V05 respectively. Use spi-tool in cli to read their id.

### CODE REVIEW

**Comment by tmedicci** [comment]
Hi @w2016561536, another tip: you did not need to delete the PR (https://github.com/apache/nuttx/pull/11553#event-11494986888) when you are asked to rebase from master. You can just rebase and force push (force pushing is totally acceptable for branches that are being written by a single person. That is the case for the  `esp32s3-spidev` branch of your own fork on NuttX).

I'll review this PR ;)

File: boards/xtensa/esp32s3/common/src/esp32s3_board_spidev.c [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: board_spidev_initialize
+ *
+ * Description:
+ *   Initialize and register SPI driver for the specified SPI port.
+ *
+ ****************************************************************************/
```
Comment (tmedicci) [inline boards/xtensa/esp32s3/common/src/esp32s3_board_spidev.c:45]: The description is missing the `Input Parameters` and `Returned Value` sections. You can refer to the following pages for an example.  

https://nuttx.apache.org/docs/latest/contributing/coding_style.html
https://nuttx.apache.org/docs/latest/contributing/coding_style.html#cfilestructure

File: boards/xtensa/esp32s3/common/include/esp32s3_board_spidev.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP32S3_SPI
+/****************************************************************************
+ * Name: board_spidev_initialize
+ *
+ * Description:
+ *   Initialize and register SPI driver for the specified SPI port.
+ *
+ ****************************************************************************/
```
Comment (tmedicci) [inline boards/xtensa/esp32s3/common/include/esp32s3_board_spidev.h:60]: The description is missing the `Input Parameters` and `Returned Value` sections. You can refer to the following pages for an example.  

https://nuttx.apache.org/docs/latest/contributing/coding_style.html
https://nuttx.apache.org/docs/latest/contributing/coding_style.html#cfilestructure

**Review by tmedicci — CHANGES_REQUESTED** [review]
Congratulations. Thanks for submitting this PR. It's very good, just some small bug fixes and improvements.

**Comment by w2016561536** [comment]
@tmedicci Thanks for your advice. I will edit it soon.

**Comment by w2016561536** [comment]
> Also, please note that the CI is failing to test [the `esp32s3-lcd-ev/ws2812` defconfig](https://github.com/apache/nuttx/actions/runs/7538371173/job/20527243651?pr=11554#step:7:874). This happens because the SPI is selected for the chip but the SPI driver on NuttX is not. You can make it a dependency in `nuttx/arch/xtensa/src/esp32s3/Kconfig`. Something like:
> 
> ```
> config ESP32S3_SPI
> 	bool
> 	select SPI_DRIVER
> 	default n
> ```
> 
> If you do that, probably the error will disappear. Please, don't forget to run `./tools/refresh.sh --silent chip:esp32s3` to make sure all defconfigs from ESP32-S3 are updated. Documentation for `refresh.sh` is [here](https://nuttx.apache.org/docs/latest/components/tools/index.html#refresh-sh) (but it doesn not show the `chip` "argument". If you are willing to help NuttX, maybe you could try to open a PR to update this doc too 😉 )

Well, if we add a `SPI_DRIVER` as the dependency of `ESP32S3_SPI`, more problems appear because `SPI_DRIVER` depends on more things like `SPI_EXCHANGE`. My idea is to add a macro `CONFIG_SPI_DRIVER` to determine whether we need to register spi bus.

**Comment by tmedicci** [comment]
> > Also, please note that the CI is failing to test [the `esp32s3-lcd-ev/ws2812` defconfig](https://github.com/apache/nuttx/actions/runs/7538371173/job/20527243651?pr=11554#step:7:874). This happens because the SPI is selected for the chip but the SPI driver on NuttX is not. You can make it a dependency in `nuttx/arch/xtensa/src/esp32s3/Kconfig`. Something like:
> > ```
> > config ESP32S3_SPI
> > 	bool
> > 	select SPI_DRIVER
> > 	default n
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If you do that, probably the error will disappear. Please, don't forget to run `./tools/refresh.sh --silent chip:esp32s3` to make sure all defconfigs from ESP32-S3 are updated. Documentation for `refresh.sh` is [here](https://nuttx.apache.org/docs/latest/components/tools/index.html#refresh-sh) (but it doesn not show the `chip` "argument". If you are willing to help NuttX, maybe you could try to open a PR to update this doc too 😉 )
> 
> Well, if we add a `SPI_DRIVER` as the dependency of `ESP32S3_SPI`, more problems appear because `SPI_DRIVER` depends on more things like `SPI_EXCHANGE`. My idea is to add a macro `CONFIG_SPI_DRIVER` to determine whether we need to register spi bus.
> 
> And build for `esp32s3-lcd-ev/ws2812` and `esp32s3-box/lvgl-3` with `-Werror`. This method can pass the build.

Forget about it: it is better to just change the config used to build `esp32s3_board_spidev.c` just like it is done on ESP32: https://github.com/apache/incubator-nuttx/blob/e11627d0fba481086ccc1e640eeadd39bfa38ccf/boards/xtensa/esp32/common/src/Make.defs#L63

In `boards/xtensa/esp32s3/common/src/Make.defs`, just change `CONFIG_ESP32S3_SPI` to `CONFIG_SPI_DRIVER`

**Comment by tmedicci** [comment]
Please add a line at the end of all the source files (except for the ones auto-generated, like the `defconfig`). According to [Coding Standard](https://nuttx.apache.org/docs/latest/contributing/coding_style.html#lines)):

> In addition, all text files should end in a single newline (\n). This avoids the “No newline at end of file” warning generated by certain tools.

**Review by tmedicci — CHANGES_REQUESTED** [review]
Please add a line at the end of all the source files (except for the ones auto-generated, like the `defconfig`). According to [Coding Standard](https://nuttx.apache.org/docs/latest/contributing/coding_style.html#lines)):

> In addition, all text files should end in a single newline (\n). This avoids the “No newline at end of file” warning generated by certain tools.

File: boards/xtensa/esp32s3/esp32s3-devkit/src/esp32s3_bringup.c [hunk 3]
```diff
@@ -138,6 +143,24 @@ int esp32s3_bringup(void)
     }
 #endif
 
+#ifdef CONFIG_ESP32S3_SPI
```
Comment (tmedicci) [inline boards/xtensa/esp32s3/esp32s3-devkit/src/esp32s3_bringup.c:146]:
```suggestion
#if defined(CONFIG_ESP32S3_SPI) && defined(CONFIG_SPI_DRIVER)
```

**Review by tmedicci — APPROVED** [review]

**Pull request merged**